### PR TITLE
coprocessor: add endpoint tests

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @qiuyesuifeng @coocood PTAL

### CODE REVIEW

File: tests/coprocessor/test_select.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    fn as_pb_table_info(&self) -> schema::TableInfo {
+        let mut tb_info = schema::TableInfo::new();
+        tb_info.set_table_id(self.t_id);
+        let mut c_info = ColumnInfo::new();
+        c_info.set_tp(TYPE_LONG as i32);
+        c_info.set_column_id(tb_info.get_table_id());
+        c_info.set_pk_handle(true);
+        tb_info.mut_columns().push(c_info);
+        for (&col_tp, &col_id) in self.c_types.iter().zip(&self.c_ids) {
+            c_info = ColumnInfo::new();
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: Why set `mut_columns` twice?

File: tests/coprocessor/test_select.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        tb_info
+    }
+
+    fn as_pb_index_info(&self, idx_off: i64) -> schema::IndexInfo {
+        let mut idx_info = schema::IndexInfo::new();
+        idx_info.set_table_id(self.t_id);
+        idx_info.set_index_id(idx_off);
+        let col_off = self.indices[idx_off as usize];
+        let mut c_info = ColumnInfo::new();
+        c_info.set_tp(self.c_types[idx_off as usize] as i32);
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: Should se use col_off here?

File: tests/coprocessor/test_select.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+fn get_row(h: i64, tbl: &TableInfo) -> Vec<(Vec<u8>, Vec<u8>)> {
+    let col_values = gen_values(h, tbl);
+    let mut kvs = vec![];
+    for (v, &id) in col_values.iter().zip(&tbl.c_ids) {
+        let (k, v) = encode_col_kv(tbl.t_id, h, id, v);
+        kvs.push((k, v));
+    }
+    for (&id, &c_id) in tbl.indices.iter().zip(&tbl.i_ids) {
+        let v = col_values[id as usize].clone();
+
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: Remove empty line.

File: tests/coprocessor/test_select.rs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+fn get_row(h: i64, tbl: &TableInfo) -> Vec<(Vec<u8>, Vec<u8>)> {
+    let col_values = gen_values(h, tbl);
+    let mut kvs = vec![];
+    for (v, &id) in col_values.iter().zip(&tbl.c_ids) {
+        let (k, v) = encode_col_kv(tbl.t_id, h, id, v);
+        kvs.push((k, v));
+    }
+    for (&id, &c_id) in tbl.indices.iter().zip(&tbl.i_ids) {
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: c_id is confused, it is better to use i_id instead.

Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: also id is for c_id_offset may be bette, i think.

Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: indices is the indexed column offsets.

File: tests/coprocessor/test_select.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    indices: Vec<i32>,
+    i_ids: Vec<i64>,
+}
+
+impl TableInfo {
+    fn as_pb_table_info(&self) -> schema::TableInfo {
+        let mut tb_info = schema::TableInfo::new();
+        tb_info.set_table_id(self.t_id);
+        let mut c_info = ColumnInfo::new();
+        c_info.set_tp(TYPE_LONG as i32);
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: As for handle is pk column, the type is TypeLongLong now.

Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: My fault, @coocood  has said that int64 is the value type, seems TypeLong/TypeLongLong are both OK.

File: tests/coprocessor/test_select.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    for row in sel_resp.get_rows() {
+        let datums = datum::decode(row.get_handle()).unwrap().0;
+        assert_eq!(datums.len(), 1);
+        if let Datum::I64(h) = datums[0] {
+            handles.push(h);
+        } else {
+            panic!("i64 expected, but got {:?}", datums[0]);
+        }
+    }
+    handles.sort();
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: Should we use sort here? Seems idx is stored in order.

File: tests/coprocessor/test_select.rs [hunk 7]
```diff
@@ -43,11 +43,11 @@ impl TableInfo {
         tb_info
     }
 
-    fn as_pb_index_info(&self, idx_off: i64) -> schema::IndexInfo {
+    fn as_pb_index_info(&self, offset: i64) -> schema::IndexInfo {
         let mut idx_info = schema::IndexInfo::new();
         idx_info.set_table_id(self.t_id);
-        idx_info.set_index_id(idx_off);
-        let col_off = self.indices[idx_off as usize];
+        idx_info.set_index_id(offset);
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: Index id should be `self.i_ids[offset as usize]`.

File: tests/coprocessor/test_select.rs [hunk 8]
```diff
@@ -20,7 +20,7 @@ struct TableInfo {
     t_id: i64,
     c_types: Vec<u8>,
     c_ids: Vec<i64>,
-    indices: Vec<i32>,
+    idx_off: Vec<i32>,
```
Comment (qiuyesuifeng) [inline tests/coprocessor/test_select.rs]: idx_offs may be better.

**Comment by BusyJay** [comment]
@ngaut @qiuyesuifeng @coocood PTAL

**Comment by qiuyesuifeng** [comment]
@BusyJay Also some comments need to reply.

File: tests/coprocessor/test_select.rs (hunk 6 above)
Comment (coocood) [inline tests/coprocessor/test_select.rs]: index is ordered by indexed column, not handle.

**Comment by BusyJay** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by BusyJay** [comment]
@ngaut @coocood PTAL

**Comment by BusyJay** [comment]
@ngaut PTAL

**Comment by ngaut** [comment]
LGTM

**Pull request merged**